README.md
#qkerntool

Nonlinear machine learning tool for classification, clustering and dimensionality reduction. It integrates 13 conditional negative definite (CND)  kernel  functions and includes the qKernel and CND kernel version of Density-based spatial clustering of applications with noise (DBSCAN), Spectral Clustering, generalized  discriminant analysis (GDA), Principal component analysis (PCA), Isomap and LLE algorithm, Sammon and tSNE.

#Examples
install.packages("qkerntool") 
library(qkerntool)


NEWS.md
*************************************************************************

Changes in version 1.18
*************************************************************************
NEW FEATURES
    o add a new argument "cndkernel" in the function "qsammon", when the class of "x" is "cndkernmatrix" 
    o add a new argument "qkernel" in the function "qsammon", when the class of "x" is "qkernmatrix" 

FIXES
    o fixed error in the function "qsammon": in feature space, we should use the square root of qkernmatrix instead of Euclidean distance 
    o fixed error messages which the function "qsammon" prints. The position of these two messages ("Warning : MaxHalves exceeded." and "Optimisation terminated - TolFun exceeded.") should be exchanged
**********************************************************

tests/testthat/testqsammon_new.R

#####################################################
data("iris")
train <- as.matrix(iris[,1:4])
labeltrain<- as.integer(iris[,5])
#######
par(mfrow = c(3,4))

kpc2 <- qsammon(train,kernel = "laplbase", qpar = list(sigma = 1000, q = 0.9), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain), main = "qsammon_laplbase:sigma=1000,q=0.9", xlab="1st Principal Component",ylab="2nd Principal Component")

kpc2 <- qsammon(train,kernel = "rbfbase", qpar = list(sigma = 2, q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain), main = "qsammon_rbfbase:sigma=2,q=0.8", xlab="1st Principal Component",ylab="2nd Principal Component")

kpc2 <- qsammon(train,kernel = "rbfbase", qpar = list(sigma = 1000, q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain), main = "qsammon_rbfbase:sigma=1000,q=0.8", xlab="1st Principal Component",ylab="2nd Principal Component")

## S4 method for signature 'matrix'
kpc2 <- qsammon(train,kernel = "rbfbase", qpar = list(sigma = 0.5, q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "nonlbase", qpar = list(alpha=0.001,q=0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "laplbase", qpar = list(sigma = 112, q = 0.9), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "ratibase", qpar = list(c = 10, q = 0.9), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "multbase", qpar = list(c = 1, q = 0.9), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "invbase", qpar = list(c = 2, q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "wavbase", qpar = list(theta=1,q=0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "powbase", qpar = list(d = 1, q = 0.9), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "logbase", qpar = list(d = 1, q = 0.9), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "caubase", qpar = list(sigma = 2,q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "chibase", qpar = list(gamma=3,q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "studbase", qpar = list(d = 2, q = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "nonlcnd", qpar = list(alpha=0.000005), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "polycnd", qpar = list(d = 1,alpha = 10, c=10), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "rbfcnd", qpar = list(gamma = 3), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "laplcnd", qpar = list(gamma = 5), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "anocnd", qpar = list(d = 1, sigma = 0.1), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "raticnd", qpar = list(c = 3), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "multcnd", qpar = list(c= 0.1), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "invcnd", qpar = list(c = 2), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "wavcnd", qpar = list(theta = 0.8), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "powcnd", qpar = list(d = 1), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "logcnd", qpar = list(d = 2), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "caucnd", qpar = list(gamma = 2), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "chicnd", qpar = list(), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "studcnd", qpar = list(d = 2), dims = 2, Initialisation = 'pca')
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "rbfbase", qpar = list(sigma = 0.5, q = 0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "nonlbase", qpar = list(alpha=0.0001,q=0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "laplbase", qpar = list(sigma = 112, q = 0.9), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "ratibase", qpar = list(c = 10, q = 0.9), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "multbase", qpar = list(c = 1, q = 0.9), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "invbase", qpar = list(c = 2, q = 0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "wavbase", qpar = list(theta=1,q=0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "powbase", qpar = list(d = 1, q = 0.9), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "logbase", qpar = list(d = 1, q = 0.9), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "caubase", qpar = list(sigma = 2,q = 0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "chibase", qpar = list(gamma=3,q = 0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "studbase", qpar = list(d = 2, q = 0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "nonlcnd", qpar = list(alpha=0.000005), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "polycnd", qpar = list(d = 1.5,alpha = 10, c=10), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "rbfcnd", qpar = list(gamma = 3), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "laplcnd", qpar = list(gamma = 5), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "anocnd", qpar = list(d = 1, sigma = 0.1), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "raticnd", qpar = list(c = 3), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "multcnd", qpar = list(c= 0.1), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "invcnd", qpar = list(c = 2), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "wavcnd", qpar = list(theta = 0.8), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "powcnd", qpar = list(d = 1), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "logcnd", qpar = list(d = 1), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "caucnd", qpar = list(gamma = 2), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "chicnd", qpar = list(), dims = 2)
plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)

kpc2 <- qsammon(train,kernel = "studcnd", qpar = list(d = 1.8), dims = 2)

plot(dimRed(kpc2), col = as.integer(labeltrain))
cndkernf(kpc2)




## S4 method for signature 'qkernmatrix'
qkfunc <- rbfbase(sigma = 0.5, q = 0.8)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- nonlbase(alpha=0.0001,q=0.8)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- laplbase(sigma = 112, q = 0.9)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- ratibase(c = 10, q = 0.9)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- multbase(c = 1, q = 0.9)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- invbase(c = 2, q = 0.8)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- powbase(d = 2, q = 0.9)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- wavbase(theta=1, q=0.9)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- logbase(d = 1, q = 0.9)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- caubase(sigma = 2,q = 0.8)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- chibase(gamma=3, q = 0.8)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))

qkfunc <- studbase(d = 2, q = 0.8)
Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, dims = 2)
plot(dimRed(kpc3), col = as.integer(labeltrain))
cndkernf(kpc3)

Ktrain2 <- qkernmatrix(qkfunc, train)
kpc3 <- qsammon(Ktrain2, k=4,dims = 2,Initialisation = 'pca')
plot(dimRed(kpc3), col = as.integer(labeltrain))
cndkernf(kpc3)

## S4 method for signature 'cndkernmatrix'
cndkfunc <- nonlcnd(alpha=0.000005)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))


cndkfunc <- polycnd(d = 1.5,alpha = 10, c=10)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- rbfcnd(gamma = 3)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- laplcnd(gamma = 5)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- anocnd(d = 1, sigma = 0.1)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- raticnd(c = 3)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- multcnd(c= 0.1)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- invcnd(c = 2)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- wavcnd(theta = 0.8)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- powcnd(d = 1)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- logcnd(d = 1)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- caucnd(gamma = 2)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- chicnd()
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndkfunc <- studcnd(d = 1.8)
cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, dims = 2)
cndkernf(kpc4)
plot(dimRed(kpc4), col = as.integer(labeltrain))

cndKtrain <- cndkernmatrix(cndkfunc, train)
kpc4 <- qsammon(cndKtrain, k=4, dims = 2,Initialisation = 'pca')
cndkernf(kpc4)
plot(dimRed(kpc4), col = as.integer(labeltrain))




R/qsammon.R
#come from version of matlab

setGeneric("qsammon",function(x, ...) standardGeneric("qsammon"))

setMethod("qsammon", signature(x = "matrix"),
          function(x, kernel = "rbfbase", qpar = list(sigma = 0.5, q = 0.9), dims = 2, Initialisation = 'random', MaxHalves = 20, MaxIter = 500, TolFun = 1e-7, na.action = na.omit, ...)
          {
            x <- na.action(x)
            x <- as.matrix(x)
            ret <- new("qsammon")
            N <- nrow(x)

            if (!is(kernel,"qkernel") && !is(kernel,"cndkernel"))
            {
              if(is(kernel,"function")) kernel <- deparse(substitute(kernel))
              kernel <- do.call(kernel, qpar)

            }
            if(!is(kernel,"qkernel") && !is(kernel,"cndkernel")) stop("kernel must inherit from class 'qkernel' or 'cndkernel'")
            if(is(kernel,"cndkernel")){
              ## Compute conditionally negative definite kernel matrix
              K <- cndkernmatrix(kernel,x)
            }
            else
              if (is(kernel,"qkernel")) K <- qkernmatrix(kernel, x)

            K <- sqrt(K)
            scales <- 0.5/sum(K)

            K <- K + diag(1,N,N)
            Dinv <- K
            Dinv=1/Dinv
            Dinv[K == 0] <- 0

            #Dinv <- 1/K
            if (Initialisation == 'pca') {
              UUDD <- svd(x)
              ydata <- UUDD$u[,1:dims] %*% diag(UUDD$d)[1:dims,1:dims]
            }else{
              ydata <- matrix(rnorm(N*dims,mean = 0,sd = 1),N,dims)
            }

            one <- matrix(1,N,dims)
            #d <- Eucdist(ydata,ydata, sEuclidean = TRUE) + diag(N)
            if(is(kernel,"cndkernel")){
              ## Compute conditionally negative definite kernel matrix
              d <- sqrt(cndkernmatrix(kernel,ydata,ydata)) + diag(N)
            }
            else
              if (is(kernel,"qkernel")) d <- sqrt(qkernmatrix(kernel, ydata,ydata)) + diag(N)

            dinv <- d
            dinv=1/dinv
            dinv[d == 0] <- 0

            delta <- K - d
            E <- sum(delta^2 * Dinv)

            #get on with it
            for (i in 1:MaxIter) {
              #compute gradient, Hessian and search direction (note it is actually
              # 1/4 of the gradient and Hessian, but the step size is just the ratio
              #of the gradient and the diagonal of the Hessian so it doesn't matter).
              delta <- dinv - Dinv
              deltaone <- delta %*% one
              g <- delta %*% ydata - ydata*deltaone
              dinv3 <- dinv^3
              ydata2 <- ydata^2
              Hdin <- dinv3 %*% ydata2 - deltaone - 2*ydata*(dinv3 %*% ydata) + ydata2*(dinv3 %*% one)

              Hdin0=Hdin
              Hdin0=1/Hdin0
              Hdin0[Hdin == 0] <- 0

              s <- -as.vector(g) * abs(as.vector(Hdin0))
              ydata_old <- ydata
              #use step-halving procedure to ensure progress is made
              #ydata <- as.vector(ydata)
              #ydata_old <- as.vector(ydata_old)

              for (j in 1:MaxHalves) {
                ydata <- ydata_old + matrix(s,nrow(ydata),ncol(ydata))

                #d <- Eucdist(ydata,ydata, sEuclidean = TRUE) + diag(N)
                if(is(kernel,"cndkernel")){
                  ## Compute conditionally negative definite kernel matrix
                  d <- cndkernmatrix(kernel,ydata, ydata)
                  d[d < 0] <- 0
                  d <- sqrt(d) + diag(N)
                }
                else
                  if (is(kernel,"qkernel")){
                    d <- qkernmatrix(kernel, ydata, ydata)
                    d[d < 0] <- 0
                    d <- sqrt(d) + diag(N)
                  }

                dinv <- d
                dinv=1/dinv
                dinv[d == 0] <- 0

                delta <- K - d
                E_new <- sum(delta^2 * Dinv)

                if (E_new < E) {
                  break
                }else{
                  s <- 0.5 * s
                }
              }
              #bomb out if too many halving steps are required
              if (j == MaxHalves) {
                message( 'Warning : MaxHalves exceeded.')
                break
              }
              # evaluate termination criterion
              if (abs((E - E_new)/E) < TolFun){
                message('Optimisation terminated - TolFun exceeded.')
                break
              }
              #report progress
              #show <- list()
              E <- E_new
              message( 'epoch =  :',i,' E = ', E*scales)
            }
            E <- E*scales

            dimRed(ret) <- ydata
            # xmatrix(ret) <- x
            kcall(ret) <- match.call()
            cndkernf(ret) <- kernel
            return(ret)
          })

#---------------------------------------------------------------------------#
setMethod("qsammon", signature(x = "cndkernmatrix"),
          function(cndkernel, x, k ,dims = 2, Initialisation = 'random', MaxHalves = 20, MaxIter = 500, TolFun = 1e-7, ...)
          {

            ret <- new("qsammon")
            N <- nrow(x)
            if(!is(x,"cndkernmatrix")) stop("x must inherit from class 'cndkernmatrix'")

            x <- sqrt(x)
            scales <- 0.5/sum(x)

            x <- x + diag(1,N,N)
            Dinv <- x
            Dinv=1/Dinv
            Dinv[x == 0] <- 0

            #Dinv <- 1/x
            if (Initialisation == 'pca') {
              loc <- cmdscale(x,k)
              UUDD <- svd(loc)
              ydata <- UUDD$u[,1:dims] %*% diag(UUDD$d)[1:dims,1:dims]
            }else{
              ydata <- matrix(rnorm(N*dims,mean = 0,sd = 1),N,dims)
            }

            one <- matrix(1,N,dims)
            #d <- Eucdist(ydata,ydata, sEuclidean = TRUE) + diag(N)

            ## Compute conditionally negative definite kernel matrix
            d <- sqrt(cndkernmatrix(cndkernel,ydata,ydata)) + diag(N)


            dinv <- d
            dinv=1/dinv
            dinv[d == 0] <- 0

            delta <- x - d
            E <- sum(delta^2 * Dinv)

            #get on with it
            for (i in 1:MaxIter) {
              #compute gradient, Hessian and search direction (note it is actually
              # 1/4 of the gradient and Hessian, but the step size is just the ratio
              #of the gradient and the diagonal of the Hessian so it doesn't matter).
              delta <- dinv - Dinv
              deltaone <- delta %*% one
              g <- delta %*% ydata - ydata*deltaone
              dinv3 <- dinv^3
              ydata2 <- ydata^2
              Hdin <- dinv3 %*% ydata2 - deltaone - 2*ydata*(dinv3 %*% ydata) + ydata2*(dinv3 %*% one)

              Hdin0=Hdin
              Hdin0=1/Hdin0
              Hdin0[Hdin == 0] <- 0

              s <- -as.vector(g) * abs(as.vector(Hdin0))
              ydata_old <- ydata
              #use step-halving procedure to ensure progress is made
              #ydata <- as.vector(ydata)
              #ydata_old <- as.vector(ydata_old)

              for (j in 1:MaxHalves) {
                ydata <- ydata_old + matrix(s,nrow(ydata),ncol(ydata))

                #d <- Eucdist(ydata,ydata, sEuclidean = TRUE) + diag(N)
                d <- cndkernmatrix(cndkernel,ydata, ydata)
                d[d < 0] <- 0
                d <- sqrt(d) + diag(N)

                dinv <- d
                dinv=1/dinv
                dinv[d == 0] <- 0

                delta <- x - d
                E_new <- sum(delta^2 * Dinv)

                if (E_new < E) {
                  break
                }else{
                  s <- 0.5 * s
                }
              }
              #bomb out if too many halving steps are required
              if (j == MaxHalves) {
                message('Warning : MaxHalves exceeded')
                break
              }
              # evaluate termination criterion
              if (abs((E - E_new)/E) < TolFun){
                message( 'Optimisation terminated - TolFun exceeded.')
                break
              }
              #report progress
              #show <- list()
              E <- E_new
              message( 'epoch =  :',i,' E = ', E*scales)
            }
            E <- E*scales

            dimRed(ret) <- ydata
            # xmatrix(ret) <- x
            kcall(ret) <- match.call()
            cndkernf(ret) <- "cndkernel"
            return(ret)
          })


#---------------------------------------------------------------------------#
setMethod("qsammon", signature(x = "qkernmatrix"),
          function(qkernel, x, k, dims = 2, Initialisation = 'random', MaxHalves = 20, MaxIter = 500, TolFun = 1e-7, ...)
          {
            ret <- new("qsammon")
            N <- nrow(x)
            if(!is(x,"qkernmatrix")) stop("x must inherit from class 'qkernmatrix'")

            x <- sqrt(x)
            scales <- 0.5/sum(x)

            x <- x + diag(1,N,N)
            Dinv <- x
            Dinv=1/Dinv
            Dinv[x == 0] <- 0

            #Dinv <- 1/x
            if (Initialisation == 'pca') {
              loc <- cmdscale(x,k)
              UUDD <- svd(loc)
              ydata <- UUDD$u[,1:dims] %*% diag(UUDD$d)[1:dims,1:dims]
            }else{
              ydata <- matrix(rnorm(N*dims,mean = 0,sd = 1),N,dims)
            }
            one <- matrix(1,N,dims)
            #d <- Eucdist(ydata,ydata, sEuclidean = TRUE) + diag(N)

            d <- sqrt(qkernmatrix(qkernel, ydata,ydata)) + diag(N)

            dinv <- d
            dinv=1/dinv
            dinv[d == 0] <- 0

            delta <- x - d
            E <- sum(delta^2 * Dinv)

            #get on with it
            for (i in 1:MaxIter) {
              #compute gradient, Hessian and search direction (note it is actually
              # 1/4 of the gradient and Hessian, but the step size is just the ratio
              #of the gradient and the diagonal of the Hessian so it doesn't matter).
              delta <- dinv - Dinv
              deltaone <- delta %*% one
              g <- delta %*% ydata - ydata*deltaone
              dinv3 <- dinv^3
              ydata2 <- ydata^2
              Hdin <- dinv3 %*% ydata2 - deltaone - 2*ydata*(dinv3 %*% ydata) + ydata2*(dinv3 %*% one)

              Hdin0=Hdin
              Hdin0=1/Hdin0
              Hdin0[Hdin == 0] <- 0

              s <- -as.vector(g) * abs(as.vector(Hdin0))
              ydata_old <- ydata
              #use step-halving procedure to ensure progress is made
              #ydata <- as.vector(ydata)
              #ydata_old <- as.vector(ydata_old)

              for (j in 1:MaxHalves) {
                ydata <- ydata_old + matrix(s,nrow(ydata),ncol(ydata))

                #d <- Eucdist(ydata,ydata, sEuclidean = TRUE) + diag(N)
                d <- qkernmatrix(qkernel, ydata, ydata)
                d[d < 0] <- 0
                d <- sqrt(d) + diag(N)

                dinv <- d
                dinv=1/dinv
                dinv[d == 0] <- 0

                #  dinv <- 1/d
                delta <- x - d
                E_new <- sum(delta^2 * Dinv)

                if (E_new < E) {
                  break
                }else{
                  s <- 0.5 * s
                }
              }
              #bomb out if too many halving steps are required
              if (j == MaxHalves) {
                message('Warning : MaxHalves exceeded.')
                break
              }
              # evaluate termination criterion
              if (abs((E - E_new)/E) < TolFun){
                message( 'Optimisation terminated - TolFun exceeded.')
                break
              }
              #report progress
              #show <- list()
              E <- E_new
              message( 'epoch =  :',i,' E = ', E*scales)
            }
            E <- E*scales

            dimRed(ret) <- ydata
            # xmatrix(ret) <- x
            kcall(ret) <- match.call()
            cndkernf(ret) <- "qkernel"
            return(ret)
          })
